crates/sui-network/src/state_sync/mod.rs

```diff
@@ -0,0 +1,854 @@
+        .map(|x| x.sequence_number().saturating_add(1))
+        .unwrap_or(0)..=checkpoint.sequence_number())
+        .map(|next| {
+            let mut peers = peers
+                .iter()
+                // Filter out any peers who can't help with this particular checkpoint
+                .filter(|(_peer_id, height)| height >= &Some(next))
+                // Filter out any peers who we aren't connected with
+                .flat_map(|(peer_id, _height)| network.peer(*peer_id))
+                .map(StateSyncClient::new)
+                .collect::<Vec<_>>();
+            rand::seq::SliceRandom::shuffle(peers.as_mut_slice(), &mut rng);
+            let peer_heights = peer_heights.clone();
+            async move {
+                if let Some(checkpoint) = peer_heights
+                    .read()
+                    .unwrap()
+                    .get_checkpoint_by_sequence_number(next)
+                {
+                    return (Some(checkpoint.to_owned()), next);
+                }
+
+                // Iterate through our selected peers trying each one in turn until we're able to
+                // successfully get the target checkpoint
+                for mut peer in peers {
+                    let request = Request::new(GetCheckpointSummaryRequest::BySequenceNumber(next))
+                        .with_timeout(DEFAULT_TIMEOUT);
+                    if let Some(checkpoint) = peer
+                        .get_checkpoint_summary(request)
+                        .await
+                        .tap_err(|e| trace!("{e:?}"))
+                        .ok()
+                        .and_then(Response::into_inner)
+                        .tap_none(|| trace!("peer unable to help sync"))
+                    {
+                        // Insert in our store in the event that things fail and we need to retry
+                        peer_heights
+                            .write()
+                            .unwrap()
+                            .insert_checkpoint(checkpoint.clone());
+                        return (Some(checkpoint), next);
+                    }
+                }
+
+                (None, next)
+            }
+        })
+        .pipe(futures::stream::iter)
+        .buffered(checkpoint_header_download_concurrency);
+
+    while let Some((maybe_checkpoint, next)) = request_stream.next().await {
+        // Verify the checkpoint
+        let checkpoint = {
+            let checkpoint = maybe_checkpoint
+                .ok_or_else(|| anyhow::anyhow!("no peers where able to help sync"))?;
+
+            if checkpoint.sequence_number() != next
+                || current.as_ref().map(|x| x.digest()) != checkpoint.previous_digest()
+            {
+                return Err(anyhow::anyhow!("detected fork"));
+            }
+
+            let committee = store.get_committee(checkpoint.epoch()).expect("should have a committee for this epoch before we began processing checkpoints for the epoch");
```

Comment:
Could a byzantine validator return a checkpoint that has an inconsistently large epoch number?

Reply:
Thanks for the note. I'll add a number of other checks prior to this line in order to weed out inconsistent epoch numbers